pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,306 @@
+  // private long _maxPinotConsumptionDelay;
+  // private long _maxPartitionId;
+
+  // Timer task to check partitions that are inactive against ideal state.
+  private TrackingTimerTask _timerTask;
+  private Timer _timer;
+
+  private ServerMetrics _serverMetrics;
+  private String _tableNameWithType;
+
+  private boolean _enableAging;
+
+  RealtimeTableDataManager _realTimeTableDataManager;
+
+  /*
+   * Helper function to update the maximum when the current maximum is removed or updated.
+   * If no samples left we set maximum to minimum so new samples can be recorded.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (long pid : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(pid);
+      if ((newMax == null) || (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  /*
+   * This function will read ideal state for the given partition and return true if ideal state reflects the
+   * partition must be hosted by the current server, false otherwise.
+   *
+   * @param partitionId Partition ID that we want to verify is still hosted by this server.
+   */
+  private boolean isPartitionHostedByThisServerPerIdealState(long partitionId) {
+    return _realTimeTableDataManager.isPartitionHostedInThisServer(partitionId);
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionId) {
+    _partitionToDelaySampleMap.remove(partitionId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForConfirmation.remove(partitionId);
+   }
+
+  // Custom Constructor
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType, RealtimeTableDataManager tdm,
+      int maxNumberOfTicksBeforeVerification, int initialTimeThreadDelayMs, int timerThreadTickIntervalMs)
+      throws RuntimeException {
+    _serverMetrics = sm;
+    _tableNameWithType = tableNameWithType;
+    _realTimeTableDataManager = tdm;
+    // TBD handle negative timer values
+    if ((maxNumberOfTicksBeforeVerification < 0) || (initialTimeThreadDelayMs < 0)
+        || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _maxNumberOfTicksBeforeVerification = maxNumberOfTicksBeforeVerification;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timerTask = new TrackingTimerTask("TimerTask" + tableNameWithType, this);
+    _timer = new Timer();
```

Comment:
Suggest using the named timer constructor, so the associated threads can be traced back to this timer.
e.g.` _timer = new Timer("ConsumptionDelayTrackerTimer");`
This will help with debuggability.

Reply:
Good idea